File: mega-stream.c
/*

  Copyright 2016 Tom Deakin, University of Bristol

  This file is part of mega-stream.

  mega-stream is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  mega-stream is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with mega-stream.  If not, see <http://www.gnu.org/licenses/>.


  This aims to investigate the limiting factor for a simple kernel, in particular
  where bandwidth limits not to be reached, and latency becomes a dominating factor.

*/

#define VERSION "0.3"

#include <float.h>
#include <omp.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

#define MIN(a,b) ((a) < (b)) ? (a) : (b)
#define MAX(a,b) ((a) > (b)) ? (a) : (b)

#define IDX2(i,j,ni) ((i)+(ni)*(j))
#define IDX3(i,j,k,ni,nj) ((i)+(ni)*IDX2((j),(k),(nj)))
#define IDX4(i,j,k,l,ni,nj,nk) ((i)+(ni)*IDX3((j),(k),(l),(nj),(nk)))
#define IDX5(i,j,k,l,m,ni,nj,nk,nl) ((i)+(ni)*IDX4((j),(k),(l),(m),(nj),(nk),(nl)))
#define IDX6(i,j,k,l,m,n,ni,nj,nk,nl,nm) ((i)+(ni)*IDX5((j),(k),(l),(m),(n),(nj),(nk),(nl),(nm)))

/*
  Arrays are defined in terms of 3 sizes: inner, middle and outer.
  The large arrays are of size inner*middle*middle*middle*outer.
  The medium arrays are of size inner*middle*middle*outer.
  The small arrays are of size inner and are indexed with 1 index.

*/
#define OUTER   64 // 3^6
#define MIDDLE  16 // 2^4
#define INNER  128 // 2^7

/* Vector length is machine-specific and should be overridden by makefile */
#ifndef VLEN
#define VLEN 8
#endif
static_assert((VLEN > 0) && ((VLEN & (VLEN-1)) == 0), "VLEN must be a power of 2.");

/* Default alignment of 2 MB page boundaries */
#define ALIGNMENT 2*1024*1024

/* Tollerance with which to check final array values */
#define TOLR 1.0E-15

/* Starting values */
#define R_START 0.0
#define Q_START 0.01
#define X_START 0.02
#define Y_START 0.03
#define Z_START 0.04
#define A_START 0.06
#define B_START 0.07
#define C_START 0.08

#ifdef __APPLE__
void* aligned_alloc(size_t alignment, size_t size)
{
    void* mem;
    posix_memalign(&mem, alignment, size);
    return mem;
}
#endif

void kernel(
  const int Ng,
  const int Ni, const int Nj, const int Nk, const int Nl, const int Nm,
  double (* restrict r)[Ng][Nl][Nk][Nj][VLEN],
  const double (* restrict q)[Ng][Nl][Nk][Nj][VLEN],
  double (* restrict x)[Ng][Nk][Nj][VLEN],
  double (* restrict y)[Ng][Nl][Nj][VLEN],
  double (* restrict z)[Ng][Nl][Nk][VLEN],
  const double (* restrict a)[VLEN],
  const double (* restrict b)[VLEN],
  const double (* restrict c)[VLEN],
  double (* restrict sum)[Nl][Nk][Nj]
);
void parse_args(int argc, char *argv[]);

/* Default strides */
int Ni = INNER;
int Nj = MIDDLE;
int Nk = MIDDLE;
int Nl = MIDDLE;
int Nm = OUTER;
int Ng;

/* Number of iterations to run benchmark */
int ntimes = 100;

int main(int argc, char *argv[])
{

  printf("MEGA-STREAM! - v%s\n\n", VERSION);

  parse_args(argc, argv);

  printf("Small arrays:  %d elements\t\t(%.1lf KB)\n",
    Ni, Ni*sizeof(double)*1.0E-3);

  printf("Medium arrays: %d x %d x %d x %d elements\t(%.1lf MB)\n",
    Ni, Nj, Nj, Nm, Ni*Nj*Nj*Nm*sizeof(double)*1.0E-6);

  printf("Large arrays:  %d x %d x %d x %d x %d elements\t(%.1lf MB)\n",
    Ni, Nj, Nj, Nj, Nm, Ni*Nj*Nj*Nj*Nm*sizeof(double)*1.0E-6);

  const double footprint = (double)sizeof(double) * 1.0E-6 * (
    2.0*Ni*Nj*Nk*Nl*Nm +  /* r, q */
    Ni*Nj*Nk*Nm +         /* x */
    Ni*Nj*Nl*Nm +         /* y */
    Ni*Nk*Nl*Nm +         /* z */
    3.0*Ni +              /* a, b, c */
    Nj*Nk*Nl*Nm           /* sum */
    );
  printf("Memory footprint: %.1lf MB\n", footprint);

  /* Total memory moved (in the best case) - the arrays plus an extra sum as update is += */
  const double moved = (double)sizeof(double) * 1.0E-6 * (
    Ni*Nj*Nk*Nl*Nm  + /* read q */
    Ni*Nj*Nk*Nl*Nm  + /* write r */
    Ni + Ni + Ni    + /* read a, b and c */
    2.0*Ni*Nj*Nk*Nm + /* read and write x */
    2.0*Ni*Nj*Nl*Nm + /* read and write y */
    2.0*Ni*Nk*Nl*Nm + /* read and write z */
    2.0*Nj*Nk*Nl*Nm   /* read and write sum */
  );

  /* Split inner-most dimension into VLEN-sized chunks */
  Ng = ((Ni + (VLEN-1)) & ~(VLEN-1)) / VLEN;
  printf("Inner dimension split into %d chunks of size %d\n", Ng, VLEN);

  printf("Running %d times\n", ntimes);

  printf("\n");

  double timings[ntimes];

  double *q = aligned_alloc(ALIGNMENT, sizeof(double)*VLEN*Nj*Nk*Nl*Nm*Ng);
  double *r = aligned_alloc(ALIGNMENT, sizeof(double)*VLEN*Nj*Nk*Nl*Nm*Ng);

  double *x = aligned_alloc(ALIGNMENT, sizeof(double)*VLEN*Nj*Nk*Nm*Ng);
  double *y = aligned_alloc(ALIGNMENT, sizeof(double)*VLEN*Nj*Nl*Nm*Ng);
  double *z = aligned_alloc(ALIGNMENT, sizeof(double)*VLEN*Nk*Nl*Nm*Ng);

  double *a = aligned_alloc(ALIGNMENT, sizeof(double)*VLEN*Ng);
  double *b = aligned_alloc(ALIGNMENT, sizeof(double)*VLEN*Ng);
  double *c = aligned_alloc(ALIGNMENT, sizeof(double)*VLEN*Ng);

  double *sum = aligned_alloc(ALIGNMENT, sizeof(double)*Nj*Nk*Nl*Nm);

  /* Initalise the data */
  #pragma omp parallel
  {
    /* q and r */
    #pragma omp for
    for (int m = 0; m < Nm; m++) {
      for (int g = 0; g < Ng; g++) {
        for (int l = 0; l < Nl; l++) {
          for (int k = 0; k < Nk; k++) {
            for (int j = 0; j < Nj; j++) {
              #pragma omp simd
              for (int v = 0; v < VLEN; v++) {
                q[IDX6(v,j,k,l,g,m,VLEN,Nj,Nk,Nl,Ng)] = Q_START;
                r[IDX6(v,j,k,l,g,m,VLEN,Nj,Nk,Nl,Ng)] = R_START;
              }
            }
          }
        }
      }
    }

    /* x */
    #pragma omp for
    for (int m = 0; m < Nm; m++) {
      for (int g = 0; g < Ng; g++) {
        for (int k = 0; k < Nk; k++) {
          for (int j = 0; j < Nj; j++) {
            #pragma omp simd
            for (int v = 0; v < VLEN; v++) {
              x[IDX5(v,j,k,g,m,VLEN,Nj,Nk,Ng)] = X_START;
            }
          }
        }
      }
    }

    /* y */
    #pragma omp for
    for (int m = 0; m < Nm; m++) {
      for (int g = 0; g < Ng; g++) {
        for (int l = 0; l < Nl; l++) {
          for (int j = 0; j < Nj; j++) {
            #pragma omp simd
            for (int v = 0; v < VLEN; v++) {
              y[IDX5(v,j,l,g,m,VLEN,Nj,Nl,Ng)] = Y_START;
            }
          }
        }
      }
    }

    /* z */
    #pragma omp for
    for (int m = 0; m < Nm; m++) {
      for (int g = 0; g < Ng; g++) {
        for (int l = 0; l < Nl; l++) {
          for (int k = 0; k < Nk; k++) {
            #pragma omp simd
            for (int v = 0; v < VLEN; v++) {
              z[IDX5(v,k,l,g,m,VLEN,Nk,Nl,Ng)] = Z_START;
            }
          }
        }
      }
    }

    /* a, b, and c */
    #pragma omp for
    for (int g = 0; g < Ng; g++) {
      #pragma omp simd
      for (int v = 0; v < VLEN; v++) {
        a[IDX2(v,g,VLEN)] = A_START;
        b[IDX2(v,g,VLEN)] = B_START;
        c[IDX2(v,g,VLEN)] = C_START;
      }
    }

    /* sum */
    #pragma omp for
    for (int m = 0; m < Nm; m++) {
      for (int l = 0; l < Nl; l++) {
        for (int k = 0; k < Nk; k++) {
          for (int j = 0; j < Nj; j++) {
            sum[IDX4(j,k,l,m,Nj,Nk,Nl)] = 0.0;
          }
        }
      }
    }
  } /* End of parallel region */

  /* Run the kernel multiple times */
  for (int t = 0; t < ntimes; t++) {
    double tick = omp_get_wtime();

    kernel(Ng,Ni,Nj,Nk,Nl,Nm,
        (double(*)[Ng][Nl][Nk][Nj][VLEN]) r,
        (const double(*)[Ng][Nl][Nk][Nj][VLEN]) q,
        (double(*)[Ng][Nk][Nj][VLEN]) x,
        (double(*)[Ng][Nl][Nj][VLEN]) y,
        (double(*)[Ng][Nl][Nk][VLEN]) z,
        (const double(*)[VLEN]) a,
        (const double(*)[VLEN]) b,
        (const double(*)[VLEN]) c,
        (double(*)[Nl][Nk][Nj]) sum);

    /* Swap the pointers */
    double *tmp = q; q = r; r = tmp;

    double tock = omp_get_wtime();
    timings[t] = tock-tick;

  }

  /* Check the results - total of the sum array */
  double total = 0.0;
  for (int i = 0; i < Nj*Nk*Nl*Nm; i++)
    total += sum[i];
  printf("Sum total: %lf\n", total);

  /* Print timings */
  double min = DBL_MAX;
  double max = 0.0;
  double avg = 0.0;
  for (int t = 1; t < ntimes; t++) {
    min = MIN(min, timings[t]);
    max = MAX(max, timings[t]);
    avg += timings[t];
  }
  avg /= (double)(ntimes - 1);

  printf("\n");
  printf("Bandwidth MB/s  Min time    Max time    Avg time\n");
  printf("%12.1f %11.6f %11.6f %11.6f\n", moved/min, min, max, avg);

  /* Free memory */
  free(q);
  free(r);
  free(x);
  free(y);
  free(z);
  free(a);
  free(b);
  free(c);
  free(sum);

  return EXIT_SUCCESS;

}

/**************************************************************************
 * Kernel
 *************************************************************************/
#include <immintrin.h>
void kernel(
  const int Ng,
  const int Ni, const int Nj, const int Nk, const int Nl, const int Nm,
  double (* restrict r)[Ng][Nl][Nk][Nj][VLEN],
  const double (* restrict q)[Ng][Nl][Nk][Nj][VLEN],
  double (* restrict x)[Ng][Nk][Nj][VLEN],
  double (* restrict y)[Ng][Nl][Nj][VLEN],
  double (* restrict z)[Ng][Nl][Nk][VLEN],
  const double (* restrict a)[VLEN],
  const double (* restrict b)[VLEN],
  const double (* restrict c)[VLEN],
  double (* restrict sum)[Nl][Nk][Nj]
  )
{
  #pragma omp parallel for
  for (int m = 0; m < Nm; m++) {
    for (int g = 0; g < Ng; g++) {
      for (int l = 0; l < Nl; l++) {
        for (int k = 0; k < Nk; k++) {
          for (int j = 0; j < Nj; j++) {
            double total = 0.0;
            _mm_prefetch((const char*) (&q[m][g][l][k][j][0] + 32*VLEN), _MM_HINT_T1);
            #pragma vector nontemporal(r)
            #pragma omp simd reduction(+:total) aligned(a,b,c,x,y,z,r,q:64)
            for (int v = 0; v < VLEN; v++) {
              /* Set r */
              r[m][g][l][k][j][v] =
                q[m][g][l][k][j][v] +
                a[g][v] * x[m][g][k][j][v] +
                b[g][v] * y[m][g][l][j][v] +
                c[g][v] * z[m][g][l][k][v];

              /* Update x, y and z */
              x[m][g][k][j][v] = 0.2*r[m][g][l][k][j][v] - x[m][g][k][j][v];
              y[m][g][l][j][v] = 0.2*r[m][g][l][k][j][v] - y[m][g][l][j][v];
              z[m][g][l][k][v] = 0.2*r[m][g][l][k][j][v] - z[m][g][l][k][v];

              /* Reduce over Ni */
              total += r[m][g][l][k][j][v];

            } /* VLEN */

            sum[m][l][k][j] += total;

          } /* Nj */
        } /* Nk */
      } /* Nl */
    } /* Ng */
  } /* Nm */
}

void parse_args(int argc, char *argv[])
{
  for (int i = 1; i < argc; i++)
  {
    if (strcmp(argv[i], "--outer") == 0)
    {
      Nm = atoi(argv[++i]);
    }
    else if (strcmp(argv[i], "--inner") == 0)
    {
      Ni = atoi(argv[++i]);
    }
    else if (strcmp(argv[i], "--middle") == 0)
    {
      int num = atoi(argv[++i]);
      Nj = num;
      Nk = num;
      Nl = num;
    }
    else if (strcmp(argv[i], "--ntimes") == 0)
    {
      ntimes = atoi(argv[++i]);
      if (ntimes < 2)
      {
        fprintf(stderr, "ntimes must be 2 or greater\n");
        exit(EXIT_FAILURE);
      }
    }
    else if (strcmp(argv[i], "--help") == 0)
    {
      printf("Usage: %s [OPTION]\n", argv[0]);
      printf("\t --outer  n \tSet size of outer dimension\n");
      printf("\t --inner  n \tSet size of middle dimensions\n");
      printf("\t --middle n \tSet size of inner dimension\n");
      printf("\t --ntimes n\tRun the benchmark n times\n");
      printf("\n");
      printf("\t Outer   is %12d elements\n", OUTER);
      printf("\t Middle are %12d elements\n", MIDDLE);
      printf("\t Inner   is %12d elements\n", INNER);
      exit(EXIT_SUCCESS);
    }
    else
    {
      fprintf(stderr, "Unrecognised argument \"%s\"\n", argv[i]);
      exit(EXIT_FAILURE);
    }
  }
}

File: mega-stream.f90
!
!
! Copyright 2016 Tom Deakin, University of Bristol
!
! This file is part of mega-stream.
!
! mega-stream is free software: you can redistribute it and/or modify
! it under the terms of the GNU General Public License as published by
! the Free Software Foundation, either version 3 of the License, or
! (at your option) any later version.
!
! mega-stream is distributed in the hope that it will be useful,
! but WITHOUT ANY WARRANTY; without even the implied warranty of
! MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
! GNU General Public License for more details.
!
! You should have received a copy of the GNU General Public License
! along with mega-stream.  If not, see <http://www.gnu.org/licenses/>.
!
! This aims to investigate the limiting factor for a simple kernel, in particular
! where bandwidth limits not to be reached, and latency becomes a dominating factor.
!

PROGRAM megastream

  USE omp_lib

  IMPLICIT NONE

  ! Constant parameters

  ! Arrays are defined in terms of 3 sizes: inner, middle and outer.
  ! The large arrays are of size inner*middle*middle*middle*outer.
  ! The medium arrays are of size inner*middle*middle*outer.
  ! The small arrays are of size inner and are indexed with 1 index.

  INTEGER, PARAMETER :: OUTER = 64
  INTEGER, PARAMETER :: MIDDLE = 16
  INTEGER, PARAMETER :: INNER = 128

  ! Tollerance with which to check final array values
  REAL(8), PARAMETER :: TOLR = 1.0E-15

  ! Starting values
  REAL(8), PARAMETER :: R_START = 0.0_8
  REAL(8), PARAMETER :: Q_START = 0.01_8
  REAL(8), PARAMETER :: X_START = 0.02_8
  REAL(8), PARAMETER :: Y_START = 0.03_8
  REAL(8), PARAMETER :: Z_START = 0.04_8
  REAL(8), PARAMETER :: A_START = 0.06_8
  REAL(8), PARAMETER :: B_START = 0.07_8
  REAL(8), PARAMETER :: C_START = 0.08_8

  ! Variables

  ! Default strides
  INTEGER :: Ni = INNER
  INTEGER :: Nj = MIDDLE
  INTEGER :: Nk = MIDDLE
  INTEGER :: Nl = MIDDLE
  INTEGER :: Nm = OUTER

  ! Number of iterations to run benchmark
  INTEGER :: ntimes = 100

  ! Arrays
  REAL(8), DIMENSION(:,:,:,:,:), POINTER :: q, r, ptr_tmp
  REAL(8), DIMENSION(:,:,:,:), ALLOCATABLE :: x, y, z
  REAL(8), DIMENSION(:), ALLOCATABLE :: a, b, c
  REAL(8), DIMENSION(:,:,:,:), ALLOCATABLE :: total

  REAL(8), DIMENSION(:), ALLOCATABLE :: timings
  REAL(8) :: tick, tock

  REAL(8) :: moved

  INTEGER :: i, j, k, l, m, t

  ! Print information
  WRITE(*, *) 'MEGA-STREAM! - v0.3'
  WRITE(*, *)
  !WRITE(*, '(a,I,a,f8.1,a)') 'Small arrays: ', S_size, ' elements', S_size*8*1.0E-3, 'KB'
  !WRITE(*, '(a,I,a,I,a,f8.1,a)') 'Medium arrays: ', S_size, 'x', M_size, ' elements', S_size*M_size*8*1.0E-6, 'MB'
  !WRITE(*, '(a,I,a,I,a,I,a,f8.1,a)') 'Large arrays: ', S_size, 'x', M_size, 'x', L_size, ' elements', S_size*M_size*L_size*8*1.0E-6, 'MB'
  !WRITE(*, '(a,f8.1,a)') 'Memory footprint: ', 8 * 1.0E-6 * ( &
    !2.0*L_size*M_size*S_size + &
    !3.0*M_size*S_size +        &
    !3.0*S_size +               &
    !L_size*M_size)             &
    !, ' MB'
  WRITE(*, *) 'Running ', ntimes, 'times'
  WRITE(*, *)


  ! Allocate memory
  NULLIFY(q, r)
  ALLOCATE(q(Ni,Nj,Nk,Nl,Nm))
  ALLOCATE(r(Ni,Nj,Nk,Nl,Nm))
  ALLOCATE(x(Ni,Nj,Nk,Nm))
  ALLOCATE(y(Ni,Nj,Nl,Nm))
  ALLOCATE(z(Ni,Nk,Nl,Nm))
  ALLOCATE(a(Ni))
  ALLOCATE(b(Ni))
  ALLOCATE(c(Ni))
  ALLOCATE(total(Nj,Nk,Nl,Nm))

  ALLOCATE(timings(ntimes))

  ! Initalise the data */
!$OMP PARALLEL
  ! q and r
!$OMP DO
  DO m = 1, Nm
    DO l = 1, Nl
      DO k = 1, Nk
        DO j = 1, Nj
          DO i = 1, Ni
            r(i,j,k,l,m) = R_START
            q(i,j,k,l,m) = Q_START
          END DO
        END DO
       END DO
     END DO
   END DO
!$OMP END DO

  ! x
!$OMP DO
  DO m = 1, Nm
    DO k = 1, Nk
      DO j = 1, Nj
        DO i = 1, Ni
          x(i,j,k,m) = X_START
        END DO
      END DO
     END DO
   END DO
!$OMP END DO

  ! y
!$OMP DO
  DO m = 1, Nm
    DO l = 1, Nl
      DO j = 1, Nj
        DO i = 1, Ni
          y(i,j,l,m) = Y_START
        END DO
      END DO
     END DO
   END DO
!$OMP END DO

  ! z
!$OMP DO
  DO m = 1, Nm
    DO l = 1, Nl
      DO k = 1, Nk
        DO i = 1, Ni
          z(i,k,l,m) = Z_START
        END DO
      END DO
     END DO
   END DO
!$OMP END DO

  ! a, b, and c
!$OMP DO
  DO i = 1, Ni
    a(i) = A_START
    b(i) = B_START
    c(i) = C_START
  END DO
!$OMP END DO

  ! sum
!$OMP DO
  DO m = 1, Nm
    DO l = 1, Nl
      DO k = 1, Nk
        DO j = 1, Nj
          total(j,k,l,m) = 1.0_8
        END DO
      END DO
    END DO
  END DO
!$OMP END DO

!$OMP END PARALLEL

  ! Run the kernel multiple times
  DO t = 1, ntimes
    tick = omp_get_wtime()

    CALL kernel(Ni, Nj, Nk, Nl, Nm, r, q, x, y, z, a, b, c, total)

    ! Swap the pointers
    ptr_tmp => q
    q => r
    r => ptr_tmp
    NULLIFY(ptr_tmp)

    tock = omp_get_wtime()
    timings(t) = tock-tick

  END DO ! t


  ! Check the results
  WRITE(*, *) "Sum total: ", SUM(total)

  ! Print timings
  WRITE(*, '(a,a,a,a)') 'Bandwidth MB/s', 'Min time', 'Max time', 'Avg time'
  WRITE(*, '(f12.1,f11.6,f11.6,f11.6)') &
    moved / MINVAL(timings(2:ntimes)), &
    MINVAL(timings(2:ntimes)), &
    MAXVAL(timings(2:ntimes)), &
    SUM(timings(2:ntimes)) / (ntimes - 1)

! Deallocate memory
  DEALLOCATE(q, r)
  DEALLOCATE(a, b, c)
  DEALLOCATE(x, y, z)
  DEALLOCATE(total)
  DEALLOCATE(timings)

END PROGRAM megastream

!**************************************************************************
!* Kernel
!*************************************************************************/
SUBROUTINE kernel(Ni, Nj, Nk, Nl, Nm, r, q, x, y, z, a, b, c, total)

  IMPLICIT NONE
  
  INTEGER, INTENT(IN) :: Ni, Nj, Nk, Nl, Nm

  REAL(8), DIMENSION(Ni, Nj, Nk, Nl, Nm), INTENT(INOUT) :: q, r
  REAL(8), DIMENSION(Ni, Nj, Nk, Nm), INTENT(INOUT) :: x
  REAL(8), DIMENSION(Ni, Nj, Nl, Nm), INTENT(INOUT) :: y
  REAL(8), DIMENSION(Ni, Nk, Nl, Nm), INTENT(INOUT) :: z
  REAL(8), DIMENSION(Ni), INTENT(IN) :: a, b, c
  REAL(8), DIMENSION(Nj, Nk, Nl, Nm), INTENT(INOUT) :: total

  ! Local variables
  INTEGER :: i, j, k, l, m
  REAL(8) :: tmp_r, tmp_total

!$OMP PARALLEL DO PRIVATE(tmp_r, tmp_total)
  DO m = 1, Nm
    DO l = 1, Nl
      DO k = 1, Nk
        DO j = 1, Nj
          tmp_total = 0.0_8
          !$OMP SIMD REDUCTION(+:tmp_total)
          DO i = 1, Ni
            ! Set r
            tmp_r = q(i,j,k,l,m) +  &
              a(i)*x(i,j,k,m) +     &
              b(i)*y(i,j,l,m) +     &
              c(i)*z(i,k,l,m)

            ! Update x, y and z
            x(i,j,k,m) = 0.2_8*tmp_r - x(i,j,k,m)
            y(i,j,l,m) = 0.2_8*tmp_r - y(i,j,l,m)
            z(i,k,l,m) = 0.2_8*tmp_r - z(i,k,l,m)

            ! Reduce over Ni
            tmp_total = tmp_total + tmp_r

            ! Save r
            r(i,j,k,l,m) = tmp_r

          END DO ! i

          total(j,k,l,m) = total(j,k,l,m) + tmp_total

        END DO ! j
      END DO ! k
    END DO ! l
  END DO ! m
!$OMP END PARALLEL DO

END SUBROUTINE kernel

